torchtune/utils/logging.py

```diff
@@ -25,3 +29,34 @@ def get_logger(level: Optional[str] = None) -> logging.Logger:
         level = getattr(logging, level.upper())
         logger.setLevel(level)
     return logger
+
+
+def deprecated(msg: str = "") -> Callable[[T], T]:
```

Comment:
To avoid blowing up people's logs, is it possible to make sure that:
1. This only logs once (not every time it is hit)
2. This only logs on rank0

Reply:
Nice-to-have: include an explicit version number when we will remove support for the API

Reply:
Also this is kinda orthogonal, but I don't like that our utils directory == trainer utils. Because it takes deps on pretty much every other directory, which means we may run into circular import issues if we try to use this in our lower-level components. I would think about whether there's a place further upstream in our dependency graph we can put this (if only we didn't already commandeer the utils folder name with our furthest downstream components..)

Reply:
1. the lru cache ensures it's logged only once per class, also verified it in the unit test
2. good point, will add this

Reply:
> there's a place further upstream in our dependency graph we can put this

not entirely sure what this would practically mean, do you mean a different file?

Reply:
Probably a separate directory. Because otherwise anywhere we add `from utils.file_name import SomeClass` will go through `utils/__init__.py`, right? And that will import all our APIs with all their upstream dependencies in data, datasets, modules, models. So we'll get stuck with a circular dependency unless we move this out of utils entirely